### PR TITLE
feat: cache handler revalidate tag functionality

### PR DESCRIPTION
**Pull request by trsergey25**
What was done:

- added revalidate tag function to NextCacheHandler
- added revalidating tests for NextCacheHandler
- added revalidate tag function to FileSystemCache
- added revalidating tests for FileSystemCache
- added revalidate tag function to MemoryCache
- added revalidating tests for MemoryCache
- added revalidate tag function to RedisCache
- added revalidating tests for RedisCache
- added revalidate tag function to S3Cache
- added revalidating tests for S3Cache
---
Additional changes:

- added expiration date to the RedisCache based on the revalidate value
- added tags to the metadata of S3Cache object

### CODE REVIEW

File: packages/redis-cache/src/index.ts [hunk 1]
```diff
@@ -9,25 +10,59 @@ export class RedisCache implements CacheStrategy {
...
+    // Revalidate by Path
+    if (tag.startsWith(NEXT_CACHE_IMPLICIT_TAG_ID)) {
+      await this.deleteAllByKeyMatch(tag.slice(NEXT_CACHE_IMPLICIT_TAG_ID.length))
+      return
+    }
+
+    // Revalidate by Tag
+    let cursor = 0
+    do {
+      const { cursor: currentCursor, keys } = await this.client.scan(0, { MATCH: '*', COUNT: 100 })
```
Comment (RomanBobrovskiy) [inline packages/redis-cache/src/index.ts:35]: in redis we can query by value https://redis.io/docs/latest/develop/interact/search-and-query/query/, so possible we can fetch all data which contains specific tag inside

File: packages/s3-cache/src/index.ts [hunk 2]
```diff
@@ -39,33 +39,97 @@ export class S3Cache implements CacheStrategy {
...
 
-    await this.client.putObject({
-      Bucket: this.bucketName,
-      Key: `${this.removeSlashFromStart(pageKey)}/${cacheKey}.json`,
-      Body: JSON.stringify(data)
-    })
+    await this.client.putObject({ ...input, Key: `${input.Key}.json`, Body: JSON.stringify(data) })
+  }
+
+  async revalidateTag(tag: string): Promise<void> {
```
Comment (RomanBobrovskiy) [inline packages/s3-cache/src/index.ts:55]: we need to think about optimizing it, i do not really like that we calling each file independently

File: packages/cache-core/__tests__/cacheHandler.spec.ts [hunk 3]
```diff
@@ -322,4 +324,60 @@ describe('CacheHandler', () => {
 
     expect(mockLogger.info).toHaveBeenCalledTimes(15)
   })
+
+  it('should log when revalidate path', async () => {
+    Cache.setCacheStrategy(new FileSystemCache())
+    Cache.setLogger(mockLogger)
+    const cacheHandler = new Cache(mockNextHandlerContext)
+
+    await cacheHandler.revalidateTag(`_N_T_${mockCacheKey}`)
```
[the remark this answers is not in the document]
Comment (trsergey25) [reply packages/cache-core/__tests__/cacheHandler.spec.ts:334]: `_N_T_` this is `NEXT_CACHE_IMPLICIT_TAG_ID` prefixt when revalidate Path

[the remark this answers is not in the document]
Comment (trsergey25) [reply packages/cache-core/__tests__/cacheHandler.spec.ts:334]: fixed

File: packages/cache-core/src/strategies/fileSystem.ts [hunk 4]
```diff
@@ -25,14 +26,56 @@ export class FileSystemCache implements CacheStrategy {
     await fs.writeFile(pathToCacheFile, JSON.stringify(data))
   }
 
+  async revalidateTag(tag: string, ctx: CacheContext): Promise<void> {
+    if (!existsSync(ctx.serverCacheDirPath)) return
+
+    // Revalidate by Path
+    if (tag.startsWith(NEXT_CACHE_IMPLICIT_TAG_ID)) {
```
[the remark this answers is not in the document]
Comment (trsergey25) [reply packages/cache-core/src/strategies/fileSystem.ts]: yes. I moved this logic to the cacheHandler

File: packages/redis-cache/src/index.ts (hunk 1 above)
Comment (trsergey25) [reply packages/redis-cache/src/index.ts:28]: This doesn't seem to be a quick change. Can I make this a separate ticket?

**Review by RomanBobrovskiy — APPROVED** [review]

**Pull request merged**